### DISCUSSION TITLE
Does exist any standardized message/way for emergency shutdown for inverters and chargers?

### DISCUSSION THREAD

**Discussion by MiyukiDark** [canboat/canboat#367] [Q&A] [answered]
Hi folks. I'm working on one electronic project and looking for a standardized way to implement an emergency shutdown for inverters and chargers. Does exist any common high-priority message to do so? I would expect so, but cannot find it in any documentation.

**Comment by keesverruijt** [comment, marked as answer]
Sorry, but the concept of "emergency shutdown" does not exist. What does exist is a "normal" way to change state of a device. The way this works is that the device that wants to control the other device uses the _report_ PGN to _command_ state using _PGN 126208 - NMEA - Command group function_.

For instance an inverter produces [PGN 127509 - Inverter Status](https://canboat.github.io/canboat/canboat.xml#pgn-127509) which contains field 5 `Inverter enable`. To ask the inverter to shut off, send it [PGN 126208 - NMEA - Command group function](https://canboat.github.io/canboat/canboat.xml#pgn-126208) (scroll down to Command group function).

So you would send PGN 126208 with the following fields:

1. Function code: 1
2. PGN: 127509
3. Priority: 7
4. Number of parameters: 1
5. Parameter: 5
6. Value: 0 (= off)

As the length of the value field is dependent on the length of the field in the target PGN, creating this is not super easy for a generalised case. You have to look up the field length in the canboat.xml file, then round up to whole bytes.

To help convert the above to canboat format to use to write to `actisense-serial`, `ikonvert-serial` etc. you can use the `command-group-function` program in CANboat to generate data to send to the bus.

For your example above, with lets say the inverter at address 30:

    $ command-group-function 30 7 127509 5=00
    2023-02-10T08:31:31.268Z,2,126208,0,30,7,01,15,f2,01,f7,01,05,00

Note that command-group-function does not have database knowledge, so it uses the number of hex digits passed on the command line. Always pass an even number of hex digits.
